google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java

```diff
@@ -0,0 +1,347 @@
+
+    toggleDockerfileSection();
+  }
+
+  @Override
+  public void disposeUIResources() {
+    // Do nothing.
+  }
+
+  @Nls
+  @Override
+  public String getDisplayName() {
+    return GctBundle.getString("appengine.flexible.facet.name");
+  }
+
+  private boolean isRuntimeCustom() {
+    return APP_ENGINE_PROJECT_SERVICE.getFlexibleRuntimeFromAppYaml(yaml.getText())
+        == FlexibleRuntime.CUSTOM;
+  }
+
+  /**
+   * Tests if a file is valid by checking whether it exists, is a regular file and is well formed.
+   */
+  private boolean isValidConfigurationFile(String path) {
+    try {
+      if (!Files.exists(Paths.get(path))
+          || !Files.isRegularFile(Paths.get(path))) {
+        return false;
+      }
+    } catch (InvalidPathException ipe) {
+      return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * Validates the configuration and turns on/off any necessary warnings.
+   */
+  private void validateConfiguration() {
+    boolean showError = false;
+
+    if (!isValidConfigurationFile(yaml.getText())) {
+      errorIcon.setVisible(true);
+      errorMessage.setVisible(true);
+      errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.yaml"));
+      showError = true;
+    }
+
+    if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
+      errorIcon.setVisible(true);
+      errorMessage.setVisible(true);
+      errorMessage.setText(GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+      showError = true;
+    }
+
+    errorIcon.setVisible(showError);
+    errorMessage.setVisible(showError);
+  }
+
+  private void toggleDockerfileSection() {
+    boolean enabled = isRuntimeCustom();
+    dockerfile.setEnabled(enabled);
+    genDockerfileButton.setEnabled(enabled);
+    // Shows no Dockerfile label if runtime is Java and yaml is valid.
+    noDockerfileLabel.setVisible(!enabled && isValidConfigurationFile(yaml.getText()));
+  }
+
+  @Override
+  public void onFacetInitialized(@NotNull Facet facet) {
+    if (facet instanceof AppEngineFlexibleFacet) {
+      ((AppEngineFlexibleFacet) facet).getConfiguration().setYamlPath(yaml.getText());
+      ((AppEngineFlexibleFacet) facet).getConfiguration().setDockerfilePath(dockerfile.getText());
+    }
+  }
+
+  /**
+   * A somewhat generic way of generating a file for a {@link TextFieldWithBrowseButton}.
+   */
+  private static class GenerateConfigActionListener implements ActionListener {
+
+    private final Project project;
+    private final String fileName;
+    private final TextFieldWithBrowseButton filePicker;
+    private final Supplier<Optional<Path>> sourceFileProvider;
+    // Used to refresh the warnings.
+    private final Runnable callback;
+
+    GenerateConfigActionListener(
+        Project project,
+        String fileName,
+        Supplier<Optional<Path>> sourceFileProvider,
+        TextFieldWithBrowseButton filePicker,
+        Runnable callback) {
```

Comment:
I didn't think a follow up was required after your last comment. After all, it could be any of those things?